test/test_tune.py

```diff
@@ -123,3 +124,44 @@ def test(self, open_db_mock: mock.Mock, close_db_mock: mock.Mock,
         self.autotune_mock.assert_called_once_with(context, 'xyz')
         assert ret1 == tuning
         assert ret2 == tuning
+        # Check that using a different device re-runs tuning
+        context.device.name = 'another device'
+        tuning2 = {'a': 3, 'b': 5}
+        self.autotune_mock.return_value = tuning2
+        ret3 = self.autotune(context, 'xyz')
+        assert ret3 == tuning2
+
+    @mock.patch('katsdpsigproc.tune._close_db')
+    @mock.patch('katsdpsigproc.tune._open_db')
+    @mock.patch('katsdpsigproc.tune.KATSDPSIGPROC_TUNE_MATCH', 'nearest')
+    def test_nearest(self, open_db_mock: mock.Mock, close_db_mock: mock.Mock,
+                     conn: sqlite3.Connection) -> None:
+        # First populate the database with one config
+        open_db_mock.return_value = conn
+        tuning = {'a': 1, 'b': 2}
+        context1 = mock.NonCallableMock()
+        context1.device.driver_version = 'mock version'
+        context1.device.platform_name = 'mock platform'
+        context1.device.name = 'mock device'
+        self.autotune_mock.return_value = tuning
+        ret = self.autotune(context1, 'xyz')
+        assert ret == tuning
+
+        # Test incremental fallback on [version | platform | device]
+        context1.device.driver_version = 'another mock version'
+        ret2 = self.autotune(context1, 'xyz')
+        context1.device.platform_name = 'another mock platform'
+        ret3 = self.autotune(context1, 'xyz')
+        context1.device.name = 'another mock device'
+        ret4 = self.autotune(context1, 'xyz')
+        assert ret == ret2 == ret3 == ret4
+
+        # Now re-run with a different config, and make sure it uses the
+        # cached value.
+        context2 = mock.NonCallableMock()
+        context2.device.driver_version = 'mock version 2'
+        context2.device.platform_name = 'mock platform 2'
+        context2.device.name = 'mock device 2'
+        self.autotune_mock.side_effect = RuntimeError
+        ret = self.autotune(context2, 'xyz')
+        assert ret == tuning
```

Comment:
You can probably delete this block - the block above it is testing the same thing but more thoroughly.